File: Data Structures & Algorithms/the-maze-iii/submission-3.cpp
class Solution {
public:
    string findShortestWay(vector<vector<int>>& maze, vector<int>& ball, vector<int>& hole) {
        // Time Complexity: O(rows * cols * log(rows * cols)) Space Complexity: O(rows * cols)
        priority_queue<tuple<int, int, int, string>, vector<tuple<int, int, int, string>>, greater<tuple<int, int, int, string>>> minHeap; // {num steps, r, c, path}
        set<tuple<int, int, string>> moves{{0,1,"r"},{1,0,"d"},{0,-1,"l"},{-1,0,"u"}};
        set<pair<int, int>> visited;
        minHeap.push({0,ball[0],ball[1], ""});
        while (!minHeap.empty()) {
            auto [steps, r, c, path] = minHeap.top();
            minHeap.pop();
            if (visited.contains({r, c})) continue;
            visited.insert({r, c});
            if (r == hole[0] && c == hole[1]) return path;
            for (auto [dr, dc, p] : moves) {
                int nr{r + dr}, nc{c + dc};
                int cur{1};
                bool hit_hole{false};
                while (min(nr, nc) >= 0 && nr < maze.size() && nc < maze[nr].size() && maze[nr][nc] != 1) {
                    if (nr == hole[0] && nc == hole[1]) {
                        hit_hole = true;
                        break;
                    }
                    nr += dr;
                    nc += dc;
                    cur++;
                }
                if (!hit_hole) {
                    nr -= dr;
                    nc -= dc;
                    cur--;
                }
                if (!visited.contains({nr, nc})) {
                    minHeap.push({steps + cur, nr, nc, path + p});
                }
            }
        }
        return "impossible";
    }
};

File: Data Structures & Algorithms/shortest-path-in-binary-matrix/submission-3.py
class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        '''
        q for nodes to explore
        visited set 
        steps = 0

        moves includes diagonal steps

        add [[0,0]] to q and start bfs layer by layer

        while q
        pop node from q
        if node == (len(grid) - 1, len(grid[0] - 1) return steps
        increment steps
        Otherwise explore surrounding nodes
        if new node is in bounds and not 1 then add to q

        return len of shortest path or -1 if goal cant be reached

        Time and Space Complexity: O(row * col) or n^2 in square grid 
        '''
        visited = set()
        q = deque()
        steps = 1
        moves = ((0,1), (1,0), (0,-1), (-1,0), (1,1), (-1,-1), (-1,1), (1,-1))
        rows = len(grid) - 1
        cols = len(grid[0]) - 1
        goal = (rows, cols)
        if grid[0][0] == 1 or grid[rows][cols] == 1:
            return -1
        q.append([0,0])
        while q:
            for i in range(len(q)):
                r, c = q.popleft()
                if (r, c) == goal: return steps
                for dr, dc in moves:
                    nr, nc = r + dr, c + dc
                    if min(nr, nc) >= 0 and nr < len(grid) and nc < len(grid[nr]) and (nr, nc) not in visited:
                        visited.add((nr, nc))
                        if grid[nr][nc] != 1: q.append([nr, nc])
            steps += 1
        return -1